raiden-ts/src/transfers/epics.ts

```diff
@@ -603,64 +663,57 @@ export const transferUnlockedRetryMessageEpic = (
  *
  * @param action$ - Observable of transferExpired actions
  * @param state$ - Observable of RaidenStates
+ * @param latest$ - RaidenEpicDeps latest
  * @returns Observable of messageSend actions
  */
 export const transferExpiredRetryMessageEpic = (
   action$: Observable<RaidenAction>,
   state$: Observable<RaidenState>,
+  { latest$ }: RaidenEpicDeps,
 ): Observable<ActionType<typeof messageSend>> =>
-  state$.pipe(
-    publishReplay(1, undefined, state$ =>
-      action$.pipe(
-        filter(isActionOf(transferExpired)),
-        withLatestFrom(state$),
-        mergeMap(([action, state]) => {
-          const secrethash = action.meta.secrethash;
-          if (!(secrethash in state.sent)) return EMPTY; // shouldn't happen
-          const lockExpired = action.payload.message,
-            send = messageSend(
-              { message: lockExpired },
-              { address: state.sent[secrethash].transfer[1].recipient },
-            );
-          // emit Send once immediatelly, then wait until respective messageSent, then completes
-          const sendOnceAndWaitSent$ = merge(
-            of(send),
-            action$.pipe(
-              filter(
-                a =>
-                  isActionOf(messageSent, a) &&
-                  a.payload.message === send.payload.message &&
-                  a.meta.address === send.meta.address,
-              ),
-              take(1),
-              // don't output messageSent, just wait for it before completing
-              ignoreElements(),
-            ),
-          );
-          return sendOnceAndWaitSent$.pipe(
-            // Resubscribe/retry every 30s after messageSend succeeds with messageSent
-            // Notice first (or any) messageSend can wait for a long time before succeeding, as it
-            // waits for address's user in transport to be online and joined room before actually
-            // sending the message. That's why repeatWhen emits/resubscribe only some time after
-            // sendOnceAndWaitSent$ completes, instead of a plain 'interval'
-            // TODO: configurable retry delay, possibly use an exponential backoff timeout strat
-            repeatWhen(completed$ => completed$.pipe(delay(30e3))),
-            // until transferExpireProcessed OR channelClosed
-            takeUntil(
-              state$.pipe(
-                filter(
-                  state =>
-                    !!state.sent[secrethash].lockExpiredProcessed ||
-                    !!state.sent[secrethash].channelClosed,
-                ),
-              ),
-            ),
-          );
-        }),
-      ),
+  action$.pipe(
+    filter(isActionOf(transferExpired)),
+    withLatestFrom(latest$.pipe(pluckDistinct('state'))),
+    mergeMap(([action, state]) =>
+      expiredRetryMessages$(action$, latest$.pipe(pluckDistinct('state')), action, state),
     ),
   );
 
+function autoExpire$(state: RaidenState, blockNumber: number, action$: Observable<RaidenAction>) {
```

Comment:
What's our strategy for documenting internal functions? Should we still add JSDoc or are inline comments enough?

JSDoc seems like a good overall strategy, because you get a quick summary of what the function does and can also export it and move it around, without having to write all the doc again. But that's just my opinion

Currently it's a bit hard to understand what `autoExpire$` does, without digging into the code

Reply:
I am not sure how we should document this extracted methods. The reason for that is that they are actually documented as methods but at the caller level. So maybe we could actually link to the actual method above with something like `core logic for ***epic`